flutter/lib/src/sentry_user_interaction_widget.dart

```diff
@@ -0,0 +1,287 @@
+
+const _tapDeltaArea = 20 * 20;
+Element? _clickTrackerElement;
+
+class SentryUserInteractionWidget extends StatefulWidget {
+  SentryUserInteractionWidget({
+    Key? key,
+    required this.child,
+    @internal Hub? hub,
+  }) : super(key: key) {
+    _hub = hub ?? HubAdapter();
+  }
+
+  final Widget child;
+
+  late final Hub _hub;
+
+  @override
+  StatefulElement createElement() {
+    final element = super.createElement();
+    _clickTrackerElement = element;
+    return element;
+  }
+
+  @override
+  _SentryUserInteractionWidgetState createState() =>
+      _SentryUserInteractionWidgetState();
+}
+
+class _SentryUserInteractionWidgetState
+    extends State<SentryUserInteractionWidget> {
+  int? _lastPointerId;
+  Offset? _lastPointerDownLocation;
+  // final Hub _hub;
+  Widget? _lastWidget;
+  ISentrySpan? _activeTransaction;
+
+  // _SentryUserInteractionWidgetState(this._hub);
+
+  @override
+  Widget build(BuildContext context) {
+    return Listener(
+      behavior: HitTestBehavior.translucent,
+      onPointerDown: _onPointerDown,
+      onPointerUp: _onPointerUp,
+      child: widget.child,
+    );
+  }
+
+  void _onPointerDown(PointerDownEvent event) {
+    _lastPointerId = event.pointer;
+    _lastPointerDownLocation = event.localPosition;
+  }
+
+  void _onPointerUp(PointerUpEvent event) {
+    // Figure out if something was tapped
+    final location = _lastPointerDownLocation;
+    if (location == null || event.pointer != _lastPointerId) {
+      return;
+    }
+    final delta = Offset(
+      location.dx - event.localPosition.dx,
+      location.dy - event.localPosition.dy,
+    );
+
+    if (delta.distanceSquared < _tapDeltaArea) {
+      // Widget was tapped
+      _onTappedAt(event.localPosition);
+    }
+  }
+
+  void _onTappedAt(Offset position) {
+    final tappedWidget = _getElementAt(position);
+    if (tappedWidget == null || tappedWidget.keyValue == null) {
+      return;
+    }
+
+    Map<String, dynamic>? data;
+    // ignore: invalid_use_of_internal_member
+    if (widget._hub.options.sendDefaultPii && tappedWidget.description.isNotEmpty) {
+      data = {};
+      data['label'] = tappedWidget.description;
+    }
+
+    const category = 'click';
+    // TODO: check if crumbs are enabled
+    final crumb = Breadcrumb.userInteraction(
+      subCategory: category,
+      viewId: tappedWidget.keyValue,
+      // viewClass: tappedWidget.element.widget.runtimeType.toString(),
+      viewClass: tappedWidget.type, // to avoid minification
+      data: data,
+    );
+    widget._hub.addBreadcrumb(crumb, hint: tappedWidget.element.widget);
+
+    // TODO: options
+    // ignore: invalid_use_of_internal_member
+    if (!widget._hub.options.isTracingEnabled()) {
+      return;
+    }
+
+    // TODO: name should be screenName.widgetName, maybe get from router?
```

Comment:
@ueman is it possible to find out which screen (maybe the root `Scaffold`?) a widget belongs to?
For example, `MainScaffold.btn_login`, I know that the `btn_login` comes from the `MainScaffold` screen.
I could not find any way that really makes sense.

Reply:
If you have the context of the widget, you can do `ModalRoute.of(context).settings.name;`.

You have to track which `StatefulElement` is the closest to the widget which gets tracked. Via that `StatefulElement`, you can do `ModalRoute.of(statefulElement.state.context).settings.name`.

So in `_getDescriptionFrom` you can do something like 
```dart
TappedWidget? _getDescriptionFrom(Element element) {
    String? currentRouteName;
    // Add the following lines
    final widget = element.widget;
    if(element is StatefulElement) {
      currentRouteName = ModalRoute.of(statefulElement.state.context).settings.name;
    }
    // Used by ElevatedButton, TextButton, OutlinedButton.
    if (widget is ButtonStyleButton) {
      if (widget.enabled) {
        return TappedWidget(
          element: element,
          description: _findDescriptionOf(element, true),
          type: 'ButtonStyleButton',
        );
      }
    } else if (widget is MaterialButton) {
      if (widget.enabled) {
        return TappedWidget(
          element: element,
          description: _findDescriptionOf(element, true),
          type: 'MaterialButton',
        );
      }
    } else if (widget is CupertinoButton) {
      if (widget.enabled) {
        return TappedWidget(
          element: element,
          description: _findDescriptionOf(element, true),
          type: 'CupertinoButton',
        );
      }
    } else if (widget is InkWell) {
      if (widget.onTap != null) {
        return TappedWidget(
          element: element,
          description: _findDescriptionOf(element, false),
          type: 'InkWell',
        );
      }
    } else if (widget is IconButton) {
      if (widget.onPressed != null) {
        return TappedWidget(
          element: element,
          description: _findDescriptionOf(element, false),
          type: 'IconButton',
        );
      }
    } else if (widget is GestureDetector) {
      if (widget.onTap != null) {
        return TappedWidget(
          element: element,
          description: '',
          type: 'GestureDetector',
        );
      }
    }

    return null;
  }
```

No clue about the performance impact of calling `ModalRoute.of` for each `StatefulElement` but it could be a bit too much. Better to optimize it :D

Reply:
Yeah and then it works only for `StatefulElement` as well, `StatelessElement` elements don't have a `state`.
Maybe I will mention in the docs for now that the `key` must be unique across screens as well.
The other option is if I make it work along with `SentryNavigatorObserver`, because `SentryNavigatorObserver` always knows what's the current `route` if the observer is installed.

Reply:
True, but I would bet that it's basically guaranteed that there's one, though.

I wouldn't use SentryNavigatorObserver tbh, because it's common to have multiple navigators and then you have to figure out which one is the navigator to get the current route from